GraphicCompression.pas

```diff
@@ -2837,6 +2842,364 @@ procedure TPCDDecoder.Encode(Source, Dest: Pointer; Count: Cardinal; var BytesSt
+  Temp: Integer;
+  Target: PByte;
+
+begin
+	// Reset decoder state from any previous strip/tile, in case application didn't read the whole strip.
+	jpeg_abort(@FState.General.Common);
+
+  FState.RawBuffer := Source;
+  FState.RawBufferSize := PackedSize;
+	// Read the header for this strip/tile.
+	jpeg_read_header(@FState.General, True);
+
+  with PImageProperties(FImageProperties)^ do
+  begin
+    // Check image parameters and set decompression parameters.
+    if ioTiled in Options then
+    begin
+      // tiled images currently not supported
+      SegmentWidth := TileWidth;
+      SegmentHeight := Height;
+      BytesPerLine := 0; //TIFFTileRowSize(tif);
+    end
+    else
+    begin
+      SegmentWidth := Width;
+      SegmentHeight := Height - CurrentRow;
+      // I assume here that all strips are equally sized
+      if SegmentHeight > RowsPerStrip[0] then SegmentHeight := RowsPerStrip[0];
+    end;
+
+    FState.BytesPerLine := BytesPerLine;
+
+    if (PlanarConfig = PLANARCONFIG_SEPARATE) and (CurrentStrip = StripCount) then
+    begin
+      // For PC 2, scale down the expected strip/tile size to match a downsampled component
+      SegmentWidth := (SegmentWidth + Cardinal(FState.HSampling - 1)) div FState.HSampling;
+      SegmentHeight := (SegmentHeight + Cardinal(FState.VSampling - 1)) div FState.VSampling;
+    end;
+
+    if (FState.General.d.image_width <> SegmentWidth) or
+       (FState.General.d.image_height <> SegmentHeight) then CompressionError(gesJPEGStripSize);
+
+    Temp := 1;
+    if PlanarConfig = PLANARCONFIG_CONTIG then Temp := SamplesPerPixel;
+    if FState.General.d.num_components <> Temp then CompressionError(gesJPEGComponentCount);
+
+    if FState.General.d.data_precision <> BitsPerSample then CompressionError(gesJPEGDataPrecision);
+
+    if PlanarConfig = PLANARCONFIG_CONTIG then
+    begin
+      // component 0 should have expected sampling factors
+      if (FState.General.d.comp_info.h_samp_factor <> FState.HSampling) or
+         (FState.General.d.comp_info.v_samp_factor <> FState.VSampling) then
+        CompressionError(gesJPEGSamplingFactors);
+
+      // rest should have sampling factors 1,1
+      for I := 1 to FState.General.d.num_components - 1 do
+        with PCompInfoArray(FState.General.d.comp_info)[I] do
+        begin
+          if (h_samp_factor <> 1) or (v_samp_factor <> 1) then
+            CompressionError(gesJPEGSamplingFactors);
+        end;
+    end
+    else
+    begin
+      // PC 2's single component should have sampling factors 1,1
+      if (FState.General.d.comp_info.h_samp_factor <> 1) or
+         (FState.General.d.comp_info.v_samp_factor <> 1) then
+        CompressionError(gesJPEGSamplingFactors);
+    end;
+
+    // Since libjpeg can convert YCbCr data to RGB (actually BGR) on the fly I let do
+    // it this conversion instead handling it by the color manager.
+    if ColorScheme = csYCbCr then FState.General.d.jpeg_color_space := JCS_YCbCr

```

Comment:
Put the parts after then and else on an own line, or you will never be able to set break points on them.